File: src/notifier.py
"""Slack通知モジュール"""
import os
from datetime import datetime
from typing import Optional

import requests


def notify_slack(message: str, level: str = "info") -> bool:
    """
    Slackに通知を送信

    Args:
        message: 通知メッセージ
        level: 通知レベル ("info", "success", "error", "warning")

    Returns:
        成功した場合True
    """
    webhook_url = os.getenv("SLACK_WEBHOOK_URL", "").strip()
    if not webhook_url:
        return False

    emoji_map = {
        "info": "ℹ️",
        "success": "✅",
        "error": "🚨",
        "warning": "⚠️",
    }
    emoji = emoji_map.get(level, "📝")

    try:
        response = requests.post(
            webhook_url,
            json={"text": f"{emoji} {message}"},
            timeout=10,
        )
        return response.status_code == 200
    except Exception as e:
        print(f"[slack] Failed to send notification: {e}")
        return False


def notify_start(keywords_count: int, day_name: str, total_keywords: int) -> bool:
    """スクレイピング開始通知"""
    now = datetime.now().strftime("%Y-%m-%d %H:%M")
    message = f"スクレイピング開始\n• 日時: {now}\n• {day_name}曜日分: {keywords_count}/{total_keywords} キーワード"
    return notify_slack(message, "info")


def notify_complete(
    keywords_count: int,
    total_records: int,
    new_records: int,
    errors: int,
    elapsed_minutes: float,
) -> bool:
    """スクレイピング完了通知"""
    now = datetime.now().strftime("%Y-%m-%d %H:%M")
    status = "success" if errors == 0 else "warning"

    lines = [
        "スクレイピング完了",
        f"• 日時: {now}",
        f"• キーワード数: {keywords_count}",
        f"• 総記録数: {total_records}",
        f"• 新規記録: {new_records}",
        f"• 所要時間: {elapsed_minutes:.1f}分",
    ]
    if errors > 0:
        lines.append(f"• エラー数: {errors}")

    return notify_slack("\n".join(lines), status)


def notify_error(error_message: str, context: Optional[str] = None) -> bool:
    """エラー通知"""
    now = datetime.now().strftime("%Y-%m-%d %H:%M")
    lines = [
        "エラー発生",
        f"• 日時: {now}",
        f"• エラー: {error_message}",
    ]
    if context:
        lines.append(f"• 詳細: {context}")

    return notify_slack("\n".join(lines), "error")


def notify_critical(error_message: str) -> bool:
    """重大エラー通知（プロセス停止など）"""
    now = datetime.now().strftime("%Y-%m-%d %H:%M")
    message = f"重大エラー - プロセス停止\n• 日時: {now}\n• エラー: {error_message}"
    return notify_slack(message, "error")

File: gas/organizer/Notifier.js
/**
 * Slack通知モジュール
 * GASからSlackへ通知を送信
 */

/**
 * Slack Webhook URLを取得
 * Script Propertiesから取得
 * @return {string} Webhook URL
 */
function getSlackWebhookUrl() {
  const props = PropertiesService.getScriptProperties();
  return props.getProperty('SLACK_WEBHOOK_URL') || '';
}

/**
 * Slack Webhook URLを設定（初回のみ手動実行）
 * @param {string} url Webhook URL
 */
function setSlackWebhookUrl(url) {
  const props = PropertiesService.getScriptProperties();
  props.setProperty('SLACK_WEBHOOK_URL', url);
  Logger.log('Slack Webhook URLを設定しました');
}

/**
 * Slackに通知を送信
 * @param {string} message メッセージ
 * @param {string} level 通知レベル (info, success, error, warning)
 * @return {boolean} 成功した場合true
 */
function notifySlack(message, level) {
  const webhookUrl = getSlackWebhookUrl();
  if (!webhookUrl) {
    Logger.log('[slack] Webhook URL not configured');
    return false;
  }

  const emojiMap = {
    'info': 'ℹ️',
    'success': '✅',
    'error': '🚨',
    'warning': '⚠️'
  };
  const emoji = emojiMap[level] || '📝';

  try {
    const payload = {
      text: `${emoji} ${message}`
    };

    const options = {
      method: 'post',
      contentType: 'application/json',
      payload: JSON.stringify(payload),
      muteHttpExceptions: true
    };

    const response = UrlFetchApp.fetch(webhookUrl, options);
    const responseCode = response.getResponseCode();

    if (responseCode === 200) {
      Logger.log('[slack] Notification sent successfully');
      return true;
    } else {
      Logger.log(`[slack] Failed with status ${responseCode}: ${response.getContentText()}`);
      return false;
    }
  } catch (error) {
    Logger.log(`[slack] Error sending notification: ${error.message}`);
    return false;
  }
}

/**
 * GAS整理処理完了通知
 * @param {number} processedCount 処理件数
 * @param {number} cleanedCount クリーニングで削除した件数
 * @param {Object} genreCounts ジャンル別件数
 */
function notifyGasComplete(processedCount, cleanedCount, genreCounts) {
  const now = Utilities.formatDate(new Date(), 'Asia/Tokyo', 'yyyy-MM-dd HH:mm');

  const lines = [
    '[GAS] データ整理完了',
    `• 日時: ${now}`,
    `• 処理件数: ${processedCount}件`,
    `• クリーニング削除: ${cleanedCount}件`
  ];

  // ジャンル別件数を追加（0件以外）
  if (genreCounts && Object.keys(genreCounts).length > 0) {
    const genreLines = Object.entries(genreCounts)
      .filter(([_, count]) => count > 0)
      .map(([genre, count]) => `  - ${genre}: ${count}件`);

    if (genreLines.length > 0) {
      lines.push('• ジャンル内訳:');
      lines.push(...genreLines);
    }
  }

  return notifySlack(lines.join('\n'), 'success');
}

/**
 * GAS処理なし通知（処理対象がなかった場合）
 */
function notifyGasNoData() {
  const now = Utilities.formatDate(new Date(), 'Asia/Tokyo', 'yyyy-MM-dd HH:mm');
  const message = `[GAS] データ整理完了\n• 日時: ${now}\n• 処理対象データなし`;
  return notifySlack(message, 'info');
}

/**
 * GASエラー通知
 * @param {string} errorMessage エラーメッセージ
 */
function notifyGasError(errorMessage) {
  const now = Utilities.formatDate(new Date(), 'Asia/Tokyo', 'yyyy-MM-dd HH:mm');
  const message = `[GAS] エラー発生\n• 日時: ${now}\n• エラー: ${errorMessage}`;
  return notifySlack(message, 'error');
}

/**
 * Slack Webhook URL設定確認とテスト通知
 * 1. プロジェクト設定 > スクリプトプロパティで SLACK_WEBHOOK_URL を設定
 * 2. この関数を実行してテスト通知を送信
 */
function setupSlackWebhook() {
  const url = getSlackWebhookUrl();
  if (!url) {
    Logger.log('エラー: SLACK_WEBHOOK_URL がスクリプトプロパティに設定されていません');
    Logger.log('プロジェクト設定 > スクリプトプロパティ で設定してください');
    return;
  }

  Logger.log('Slack Webhook URL: 設定済み');

  // テスト通知
  const result = notifySlack('GASからのテスト通知です（セットアップ完了）', 'success');
  Logger.log(`Test notification result: ${result}`);
}

/**
 * テスト通知を送信
 */
function testSlackNotification() {
  const result = notifySlack('GASからのテスト通知です', 'info');
  Logger.log(`Test notification result: ${result}`);
}
